

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Practical 10 &mdash; Jupman</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
        <script type="text/javascript" src="_static/js/jupman.js"></script>
        <script type="text/javascript" src="_static/js/pytutor-embed.bundle.min.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/jupman.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/jupman-web.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="toc-page.html" class="icon icon-home"> Jupman
          

          
          </a>

          
            
            
              <div class="version">
                dev
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">Scientific Programming for QCB</a></li>
<li class="toctree-l1"><a class="reference internal" href="index.html#General-Info">General Info</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#Timetable-and-lecture-rooms">Timetable and lecture rooms</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#Moodle">Moodle</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#Zoom-links">Zoom links</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#Slides">Slides</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#Acknowledgements">Acknowledgements</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Practical 1</a><ul>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#Slides">Slides</a></li>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#Setting-up-the-environment">Setting up the environment</a><ul>
<li class="toctree-l3"><a class="reference internal" href="introduction.html#Linux-on-windows">Linux on windows</a></li>
<li class="toctree-l3"><a class="reference internal" href="introduction.html#A-dual-boot-system">A dual boot system</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#Our-toolbox">Our toolbox</a></li>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#Installing-Python3-in-Linux">Installing Python3 in Linux</a></li>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#Installing-Python3-in-Windows/Mac">Installing Python3 in Windows/Mac</a><ul>
<li class="toctree-l3"><a class="reference internal" href="introduction.html#OPTION-1:">OPTION 1:</a></li>
<li class="toctree-l3"><a class="reference internal" href="introduction.html#OPTION-2-(easier):">OPTION 2 (easier):</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#The-console">The console</a></li>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#Visual-Studio-Code">Visual Studio Code</a></li>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#The-debugger">The debugger</a></li>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#A-quick-Jupyter-primer-(just-for-your-information,-skip-if-not-interested)">A quick Jupyter primer (just for your information, skip if not interested)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="introduction.html#Installation">Installation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="introduction.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="practical2.html">Practical 2</a><ul>
<li class="toctree-l2"><a class="reference internal" href="practical2.html#Slides">Slides</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical2.html#Modules">Modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical2.html#Objects">Objects</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical2.html#Variables">Variables</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical2.html#Numeric-types">Numeric types</a><ul>
<li class="toctree-l3"><a class="reference internal" href="practical2.html#Integers">Integers</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical2.html#Booleans">Booleans</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical2.html#Real-numbers">Real numbers</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="practical2.html#Strings">Strings</a><ul>
<li class="toctree-l3"><a class="reference internal" href="practical2.html#Methods-for-the-str-object">Methods for the str object</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="practical2.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="practical3.html">Practical 3</a><ul>
<li class="toctree-l2"><a class="reference internal" href="practical3.html#Slides">Slides</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical3.html#Lists">Lists</a><ul>
<li class="toctree-l3"><a class="reference internal" href="practical3.html#Operators-for-lists">Operators for lists</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical3.html#Methods-of-the-class-list">Methods of the class list</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical3.html#From-strings-to-lists,-the-split-method">From strings to lists, the <code class="docutils literal notranslate"><span class="pre">split</span></code> method</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical3.html#And-back-to-strings-with-the-join-method">And back to strings with the <code class="docutils literal notranslate"><span class="pre">join</span></code> method</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="practical3.html#Tuples">Tuples</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical3.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="practical4.html">Practical 4</a><ul>
<li class="toctree-l2"><a class="reference internal" href="practical4.html#Slides">Slides</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical4.html#Execution-flow">Execution flow</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical4.html#Conditionals">Conditionals</a><ul>
<li class="toctree-l3"><a class="reference internal" href="practical4.html#The-basic-if---else-statement">The basic <em>if - else</em> statement</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical4.html#The-if---elif---else-statement">The <em>if - elif - else</em> statement</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="practical4.html#Loops">Loops</a><ul>
<li class="toctree-l3"><a class="reference internal" href="practical4.html#For-loop">For loop</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical4.html#Looping-over-a-range">Looping over a range</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical4.html#Nested-for-loops">Nested for loops</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical4.html#While-loops">While loops</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="practical4.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="practical5.html">Practical 5</a><ul>
<li class="toctree-l2"><a class="reference internal" href="practical5.html#Slides">Slides</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical5.html#More-on-loops">More on loops</a><ul>
<li class="toctree-l3"><a class="reference internal" href="practical5.html#Ternary-operator">Ternary operator</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical5.html#Break-and-continue">Break and continue</a><ul>
<li class="toctree-l4"><a class="reference internal" href="practical5.html#Continue-statement">Continue statement</a></li>
<li class="toctree-l4"><a class="reference internal" href="practical5.html#Break-statement">Break statement</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="practical5.html#List-comprehension">List comprehension</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="practical5.html#Dictionaries">Dictionaries</a><ul>
<li class="toctree-l3"><a class="reference internal" href="practical5.html#Functions-working-on-dictionaries">Functions working on dictionaries</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical5.html#Dictionary-methods">Dictionary methods</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="practical5.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="practical6.html">Practical 6</a><ul>
<li class="toctree-l2"><a class="reference internal" href="practical6.html#Slides">Slides</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical6.html#Functions">Functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical6.html#Namespace-and-variable-scope">Namespace and variable scope</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical6.html#Argument-passing">Argument passing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="practical6.html#Positional-arguments">Positional arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical6.html#Passing-arguments-by-keyword">Passing arguments by keyword</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical6.html#Specifying-default-values">Specifying default values</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="practical6.html#File-input-and-output">File input and output</a><ul>
<li class="toctree-l3"><a class="reference internal" href="practical6.html#Opening-a-file">Opening a file</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical6.html#Closing-a-file">Closing a file</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical6.html#Reading-from-a-file">Reading from a file</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical6.html#Writing-to-a-file">Writing to a file</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical6.html#String-formatting-with-format">String formatting with format</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="practical6.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="practical7.html">Practical 7</a><ul>
<li class="toctree-l2"><a class="reference internal" href="practical7.html#Slides">Slides</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical7.html#Functions">Functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical7.html#Getting-input-from-the-command-line">Getting input from the command line</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical7.html#Argparse">Argparse</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical7.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="practical8.html">Practical 8</a><ul>
<li class="toctree-l2"><a class="reference internal" href="practical8.html#Slides">Slides</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical8.html#Libraries-installation">Libraries installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical8.html#Pandas">Pandas</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical8.html#Series">Series</a><ul>
<li class="toctree-l3"><a class="reference internal" href="practical8.html#How-to-define-and-access-a-Series">How to define and access a Series</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical8.html#Operator-broadcasting">Operator broadcasting</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical8.html#Filtering">Filtering</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical8.html#Missing-data">Missing data</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical8.html#Computing-stats">Computing stats</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="practical8.html#Plotting-data">Plotting data</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical8.html#Pandas-DataFrames">Pandas DataFrames</a><ul>
<li class="toctree-l3"><a class="reference internal" href="practical8.html#Define-a-DataFrame">Define a DataFrame</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical8.html#Loading-data-from-external-files">Loading data from external files</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical8.html#Extract-values-by-row-and-column">Extract values by row and column</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical8.html#Broadcasting,-filtering-and-computing-stats">Broadcasting, filtering and computing stats</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical8.html#Merging-DataFrames">Merging DataFrames</a></li>
<li class="toctree-l3"><a class="reference internal" href="practical8.html#Grouping">Grouping</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="practical8.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="practical9.html">Practical 9</a><ul>
<li class="toctree-l2"><a class="reference internal" href="practical9.html#Slides">Slides</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical9.html#Numpy">Numpy</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical9.html#Numpy-ndarray">Numpy ndarray</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical9.html#Random-arrays">Random arrays</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical9.html#Numpy-to-and-from-pandas">Numpy to and from pandas</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical9.html#Reshaping">Reshaping</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical9.html#Iterating-over-arrays-and-Indexing">Iterating over arrays and Indexing</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical9.html#Broadcasting-and-arithmetic-functions">Broadcasting and arithmetic functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical9.html#Filtering">Filtering</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical9.html#Linear-algebra">Linear algebra</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical9.html#Matplotlib">Matplotlib</a></li>
<li class="toctree-l2"><a class="reference internal" href="practical9.html#Exercises">Exercises</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="toc.html"> </a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="toc-page.html">Jupman</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          


    <div id='jupman-langs'></div>

     <!-- <li> <h2 style="background-color:red; color:yellow">WARNING! THIS VERSION IS OLD !  <br/>
    <a style=" text-decoration: underline;" href="https://jupman.readthedocs.io/en/latest">GOT TO LATEST VERSION</a></li>
 </h2></li>-->
    <!-- Commented because doesn't work with sections,
         see https://github.com/DavidLeoni/jupman/issues/39
    <li><a href="toc-page.html">Docs</a> &raquo;</li>
    
    <li>Practical 10</li>
    -->

          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Practical-10">
<h1>Practical 10<a class="headerlink" href="#Practical-10" title="Permalink to this headline">¶</a></h1>
<p>In this practical we will will get acquainted with Biopython.</p>
<div class="section" id="Slides">
<h2>Slides<a class="headerlink" href="#Slides" title="Permalink to this headline">¶</a></h2>
<p>The slides of the introduction can be found here: <a class="reference external" href="docs/Practical10.pdf">Intro</a></p>
</div>
<div class="section" id="Biopython">
<h2>Biopython<a class="headerlink" href="#Biopython" title="Permalink to this headline">¶</a></h2>
<p>From the Biopython tutorial: The Biopython Project is an international association of developers of freely available Python tools for computational molecular biology.</p>
<p>The goal of Biopython is <strong>to make it as easy as possible to use Python for bioinformatics by creating high-quality, reusable modules and classes.</strong> Biopython features include parsers for various Bioinformatics file formats (FASTA, FASTQ, BLAST, PDB, Clustalw, Genbank,…), access to online services (NCBI, Expasy,…), interfaces to common and not-so-common programs (Clustalw, DSSP, MSMS…), a standard sequence class, various clustering modules, a KD tree data structure etc. and even some
documentation :-).</p>
<p>In this practical we will see some features of Biopython but please refer to <a class="reference external" href="http://biopython.org/wiki/Documentation">biopython documentation</a> to discover all its features, recipes etc.</p>
<p>These notes are largely based on what available <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.pdf">here</a>.</p>
</div>
<div class="section" id="Installation-of-Biopython">
<h2>Installation of Biopython<a class="headerlink" href="#Installation-of-Biopython" title="Permalink to this headline">¶</a></h2>
<p>To test if Biopython is installed in your system you can import the library with:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>import Bio
</pre></div>
</div>
<p>if the execution gives:</p>
<p><img alt="imp1" src="_images/import.png" /></p>
<p>Biopython needs to be installed.</p>
<p><strong>In windows</strong> installing Biopython should be as easy as opening the command prompt as admininstrator (typing <code class="docutils literal notranslate"><span class="pre">cmd</span></code> and then right clicking on the link choosing run as admininstrator) and then <code class="docutils literal notranslate"><span class="pre">pip3</span> <span class="pre">install</span> <span class="pre">biopython</span></code>. If that does not work, you should be able to install biopython with the following: <code class="docutils literal notranslate"><span class="pre">python3.xx</span> <span class="pre">-m</span> <span class="pre">pip</span> <span class="pre">install</span> <span class="pre">biopython</span></code> (where you should change python3.xx with the path to your python3.xx executable).</p>
<p><strong>In linux</strong> <code class="docutils literal notranslate"><span class="pre">sudo</span> <span class="pre">pip3</span> <span class="pre">install</span> <span class="pre">biopython</span></code> will install biopython for python3 up to python3.5. On python 3.6 and later the command is: <code class="docutils literal notranslate"><span class="pre">python3.6</span> <span class="pre">-m</span> <span class="pre">pip</span> <span class="pre">install</span> <span class="pre">biopython</span></code> or, more in general, <code class="docutils literal notranslate"><span class="pre">pythonX.X</span> <span class="pre">-m</span> <span class="pre">pip</span> <span class="pre">install</span> <span class="pre">biopython</span></code>.</p>
<p>Detailed installation instructions can be found <a class="reference external" href="https://github.com/biopython/biopython">here</a>.</p>
</div>
<div class="section" id="General-overview">
<h2>General overview<a class="headerlink" href="#General-overview" title="Permalink to this headline">¶</a></h2>
<p>Biopython provides the capability of parsing standard bioinformatics formats into python usable data structures. Some of the supported tools and formats are:</p>
<ol class="arabic simple">
<li><p>Blast output – both from standalone and WWW Blast</p></li>
<li><p>Clustalw</p></li>
<li><p>FASTA</p></li>
<li><p>GenBank</p></li>
<li><p>PubMed and Medline</p></li>
<li><p>ExPASy files, like Enzyme and Prosite</p></li>
<li><p>SCOP, including ‘dom’ and ‘lin’ files</p></li>
<li><p>UniGene</p></li>
<li><p>SwissProt</p></li>
</ol>
<p>among the others. In particular it provides a sequence class to perform the most common operations on sequences, parsing of alignments, clustering etc. and it can quickly connect to the most popular databases to download data or parse information on the fly.</p>
</div>
<div class="section" id="Sequence-objects">
<h2>Sequence objects<a class="headerlink" href="#Sequence-objects" title="Permalink to this headline">¶</a></h2>
<p>Biopython has a specific object <code class="docutils literal notranslate"><span class="pre">Seq</span></code> to deal with biological sequences which is a lot more powerful than the <code class="docutils literal notranslate"><span class="pre">str</span></code> object of python when it comes to manage biological data. The object <strong>Seq</strong> has methods like <strong>translate()</strong> and <strong>reverse_complement()</strong> which are very handy.</p>
<p>As in the case of python’s <code class="docutils literal notranslate"><span class="pre">str</span></code> objects, the object <code class="docutils literal notranslate"><span class="pre">Seq</span></code> is immutable. The mutable version is <code class="docutils literal notranslate"><span class="pre">MutableSeq</span></code>.</p>
<p>All the information on Seq objects can be found <a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.Seq.Seq-class.html">here</a>.</p>
<div class="section" id="The-alphabet">
<h3>The alphabet<a class="headerlink" href="#The-alphabet" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Seq</span></code> objects are composed by two information: the actual sequence and the alphabet the sequence refers to. We normally will use the <a class="reference external" href="https://www.bioinformatics.org/sms/iupac.html">IUPAC</a> alphabet as defined in the <code class="docutils literal notranslate"><span class="pre">Bio.Alphabet.IUPAC</span></code> which provides a suitable alphabet to deal with DNA, RNA and proteins. Several options are available, including <code class="docutils literal notranslate"><span class="pre">Bio.Alphabet.generic_dna</span></code> or <code class="docutils literal notranslate"><span class="pre">Bio.Alphabet.generic_protein</span></code>, <code class="docutils literal notranslate"><span class="pre">Bio.Alphabet.ThreeLetterProtein</span></code> and other.</p>
<p>Specifying an alphabet quickly tells you what information you are dealing with and moreover it provides a way to specify a constraint on the data.</p>
<p>If the definition of the alphabet is not specified, the default one is used (<code class="docutils literal notranslate"><span class="pre">Alphabet</span></code>).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">IUPAC</span>

<span class="c1">#No alphabet specified</span>
<span class="n">s</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GATTACATAATA&quot;</span><span class="p">)</span>
<span class="n">dna_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GATTATACGTAC&quot;</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">unambiguous_dna</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;S:&quot;</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;S&#39;s alphabet:&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">.</span><span class="n">alphabet</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;dna_seq:&quot;</span><span class="p">,</span> <span class="n">dna_seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;dna_seq&#39;s alphabet:&quot;</span><span class="p">,</span> <span class="n">dna_seq</span><span class="o">.</span><span class="n">alphabet</span><span class="p">)</span>

<span class="n">my_prot</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MGNAAAAKKGSEQE&quot;</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;my_prot:&quot;</span><span class="p">,</span> <span class="n">my_prot</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;my_prot&#39;s alphabet:&quot;</span><span class="p">,</span> <span class="n">my_prot</span><span class="o">.</span><span class="n">alphabet</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">ImportError</span>                               Traceback (most recent call last)
<span class="ansi-green-fg">&lt;ipython-input-1-b3b0a618da62&gt;</span> in <span class="ansi-cyan-fg">&lt;module&gt;</span>
<span class="ansi-green-intense-fg ansi-bold">      1</span> <span class="ansi-green-fg">from</span> Bio<span class="ansi-blue-fg">.</span>Seq <span class="ansi-green-fg">import</span> Seq
<span class="ansi-green-fg">----&gt; 2</span><span class="ansi-red-fg"> </span><span class="ansi-green-fg">from</span> Bio<span class="ansi-blue-fg">.</span>Alphabet <span class="ansi-green-fg">import</span> IUPAC
<span class="ansi-green-intense-fg ansi-bold">      3</span>
<span class="ansi-green-intense-fg ansi-bold">      4</span> <span class="ansi-red-fg">#No alphabet specified</span>
<span class="ansi-green-intense-fg ansi-bold">      5</span> s <span class="ansi-blue-fg">=</span> Seq<span class="ansi-blue-fg">(</span><span class="ansi-blue-fg">&#34;GATTACATAATA&#34;</span><span class="ansi-blue-fg">)</span>

<span class="ansi-green-fg">~/.local/lib/python3.6/site-packages/Bio/Alphabet/__init__.py</span> in <span class="ansi-cyan-fg">&lt;module&gt;</span>
<span class="ansi-green-intense-fg ansi-bold">     19</span>
<span class="ansi-green-intense-fg ansi-bold">     20</span> raise ImportError(
<span class="ansi-green-fg">---&gt; 21</span><span class="ansi-red-fg">     </span><span class="ansi-blue-fg">&#34;Bio.Alphabet has been removed from Biopython. In many cases, the alphabet can simply be ignored and removed from scripts. In a few cases, you may need to specify the ``molecule_type`` as an annotation on a SeqRecord for your script to work correctly. Please see https://biopython.org/wiki/Alphabet for more information.&#34;</span>
<span class="ansi-green-intense-fg ansi-bold">     22</span> )

<span class="ansi-red-fg">ImportError</span>: Bio.Alphabet has been removed from Biopython. In many cases, the alphabet can simply be ignored and removed from scripts. In a few cases, you may need to specify the ``molecule_type`` as an annotation on a SeqRecord for your script to work correctly. Please see https://biopython.org/wiki/Alphabet for more information.
</pre></div></div>
</div>
</div>
</div>
<div class="section" id="Looping,-slicing-and-concatenation">
<h2>Looping, slicing and concatenation<a class="headerlink" href="#Looping,-slicing-and-concatenation" title="Permalink to this headline">¶</a></h2>
<p>Note also that <strong>Seq objects behave like strings</strong>, but the consistency of the alphabet is checked too. For example we cannot concatenate a <strong>unambiguous_dna</strong> with a <strong>IUPAC.protein</strong> sequence. This is the error message we would get:</p>
<p><img alt="mess" src="_images/mess.png" /></p>
<p>To do it one should change the alphabet to <code class="docutils literal notranslate"><span class="pre">generic_alphabet</span></code>:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">generic_alphabet</span>

<span class="n">dna_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GATTATACGTAC&quot;</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">unambiguous_dna</span><span class="p">)</span>
<span class="n">my_prot</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MGNAAAAKKGSEQE&quot;</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">protein</span><span class="p">)</span>

<span class="n">my_prot</span><span class="o">.</span><span class="n">alphabet</span> <span class="o">=</span> <span class="n">generic_alphabet</span>

<span class="c1">#Does it really make sense though?!?</span>
<span class="nb">print</span><span class="p">(</span><span class="n">dna_seq</span> <span class="o">+</span> <span class="n">my_prot</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>We can also loop through the elements and slice strings with the usual [S:E:step] operator (and as usual S is included, E is excluded). It is also possible to convert a Seq object into a string.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">IUPAC</span>

<span class="n">dna_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GATTATACGTACGGCTA&quot;</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">unambiguous_dna</span><span class="p">)</span>

<span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">dna_seq</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">base</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

<span class="n">sub_seq</span> <span class="o">=</span> <span class="n">dna_seq</span><span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">10</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">sub_seq</span><span class="p">)</span>

<span class="c1">#Let&#39;s reverse the string:</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Reversed: &quot;</span><span class="p">,</span> <span class="n">dna_seq</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
<span class="c1">#from Seq to string:</span>
<span class="n">dna_str</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">dna_seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;As string:&quot;</span><span class="p">,</span> <span class="n">dna_str</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">dna_str</span><span class="p">))</span>
</pre></div>
</div>
</div>
<p>The object Seq provides several methods similar to the corresponding methods of strings:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.count(s)</span></code> : counts the number of times s appears in the sequence;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.upper()</span></code> : makes the sequence of the object Seq upper case</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.lower()</span></code> : makes the sequence of the object Seq lower case</p></li>
</ol>
<p><strong>Example:</strong> Let’s compute the GC content of the sequence “GATTRWWACGTACGGCTASATTACSCCGGCTA”.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">IUPAC</span>
<span class="kn">from</span> <span class="nn">Bio.SeqUtils</span> <span class="kn">import</span> <span class="n">GC</span>

<span class="n">dna_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GATTRWWACGTACGGCTASATTACSCCGGCTA&quot;</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">ambiguous_dna</span><span class="p">)</span>

<span class="n">gc</span> <span class="o">=</span> <span class="p">(</span><span class="n">dna_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;G&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">dna_seq</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;C&quot;</span><span class="p">))</span><span class="o">/</span><span class="nb">len</span><span class="p">(</span><span class="n">dna_seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;GC </span><span class="si">% i</span><span class="s2">s </span><span class="si">{:.2f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="mi">100</span><span class="o">*</span><span class="n">gc</span><span class="p">))</span>

<span class="n">correct_gc</span> <span class="o">=</span> <span class="n">GC</span><span class="p">(</span><span class="n">dna_seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;GC </span><span class="si">% i</span><span class="s2">s </span><span class="si">{:.2f}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">correct_gc</span><span class="p">))</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;original:&quot;</span><span class="p">,</span> <span class="n">dna_seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;lower_case:&quot;</span><span class="p">,</span><span class="n">dna_seq</span><span class="o">.</span><span class="n">lower</span><span class="p">())</span>
</pre></div>
</div>
</div>
<p>The difference in the results goes down to the fact that Biopython’s GC method takes into account ambiguous nucleotides like (S which stands for C or G <a class="reference external" href="https://www.bioinformatics.org/sms/iupac.html">ref iupac</a>). This is a method provided in the <strong>SeqUtils</strong> module (more info on all the methods <a class="reference external" href="http://biopython.org/DIST/docs/api/Bio.SeqUtils-module.html">here</a>).</p>
<div class="section" id="Complement-and-reverse-complement">
<h3>Complement and reverse complement<a class="headerlink" href="#Complement-and-reverse-complement" title="Permalink to this headline">¶</a></h3>
<p>A usual operation on DNA strings is the reverse complement. Biopython provides two specific methods to do that:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.complement()</span></code> to complement the sequence</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.reverse_complement()</span></code> to reverse complement the sequence.</p></li>
</ol>
<p>Note that these methods will fail with a <code class="docutils literal notranslate"><span class="pre">Value</span> <span class="pre">Error</span></code> if we try to apply them to something that is not a nucleotide sequence (e.g. proteins).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">IUPAC</span>

<span class="n">my_seq</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GATCGATGGGCCTATATAGGATCGAAAATCGC&quot;</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">unambiguous_dna</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sequence (alphabet: </span><span class="si">{}</span><span class="s2">):&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">my_seq</span><span class="o">.</span><span class="n">alphabet</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">my_seq</span><span class="p">)</span>
<span class="n">comp</span>  <span class="o">=</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">complement</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Complement (alphabet: </span><span class="si">{}</span><span class="s2">):&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">comp</span><span class="o">.</span><span class="n">alphabet</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">comp</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">revcomp</span> <span class="o">=</span> <span class="n">my_seq</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Rev Complement (alphabet: </span><span class="si">{}</span><span class="s2">):&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">revcomp</span><span class="o">.</span><span class="n">alphabet</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">revcomp</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Transcription-and-translation">
<h3>Transcription and translation<a class="headerlink" href="#Transcription-and-translation" title="Permalink to this headline">¶</a></h3>
<p>Remember that:</p>
<p><img alt="tt" src="_images/trans_transl.png" /></p>
<p>[taken from the Biopython tutorial].</p>
<p>Biopython has a method to transcribe a DNA sequence into a mRNA sequence (and one to go back):</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.transcribe()</span></code> that transcribes a DNA sequence into mRNA (converting Ts in Us).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.back_transcribe()</span></code> that back-transcribes a mRNA sequence into DNA (converting Us in Ts).</p></li>
</ol>
<p>The alphabet of the transcribed sequence is changed to RNA. Note that transcribing a protein will result in a <strong>ValueError</strong>.</p>
<p>Alike, a DNA or mRNA sequence can be translated into a protein with:</p>
<p><code class="docutils literal notranslate"><span class="pre">Seq.translate(table=N,</span> <span class="pre">to_stop=True/False,</span> <span class="pre">stop_symbol</span> <span class="pre">=symbol)</span></code> that converts a DNA sequence or mRNA into the corresponding protein.</p>
<p>Optionally, the translate method accepts a parameter <code class="docutils literal notranslate"><span class="pre">table</span></code> to specify the NCBI translation table used (see below for codes). <code class="docutils literal notranslate"><span class="pre">to_stop</span></code> makes the translation stop at the first stop codon if True (<strong>in this case the stop codon is not translated!</strong>), <code class="docutils literal notranslate"><span class="pre">stop_symbol</span></code> to replace the asterisk with any other stop character (Es. “&#64;”).</p>
<p><img alt="ncbitab" src="_images/ncbi_table.png" /> Stop codons will be represented by an asterisk.</p>
<p><strong>Example:</strong> Let’s translate into mRNA the coding sequence: ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">IUPAC</span>

<span class="n">coding_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">,</span>
                 <span class="n">IUPAC</span><span class="o">.</span><span class="n">unambiguous_dna</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">coding_dna</span><span class="p">)</span>

<span class="n">mrna</span> <span class="o">=</span> <span class="n">coding_dna</span><span class="o">.</span><span class="n">transcribe</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mrna</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mrna</span><span class="o">.</span><span class="n">alphabet</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;... and back&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mrna</span><span class="o">.</span><span class="n">back_transcribe</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Translation to protein:&quot;</span><span class="p">)</span>
<span class="n">prot</span> <span class="o">=</span> <span class="n">mrna</span><span class="o">.</span><span class="n">translate</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">prot</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Up to first stop:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mrna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">to_stop</span> <span class="o">=</span> <span class="kc">True</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Mitocondrial translation: (TGA is W!)&quot;</span><span class="p">)</span>
<span class="n">mit_prot</span> <span class="o">=</span> <span class="n">mrna</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span><span class="n">table</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mit_prot</span><span class="p">)</span>

</pre></div>
</div>
</div>
<p>Translation tables can be accessed at runtime through:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Data</span> <span class="kn">import</span> <span class="n">CodonTable</span>
<span class="c1">#equivalent to:</span>
<span class="c1">#standard_table = CodonTable.unambiguous_dna_by_id[1]</span>
<span class="n">standard_table</span> <span class="o">=</span> <span class="n">CodonTable</span><span class="o">.</span><span class="n">unambiguous_dna_by_name</span><span class="p">[</span><span class="s2">&quot;Standard&quot;</span><span class="p">]</span>
<span class="c1">#equivalent to:</span>
<span class="c1">#standard_table = CodonTable.unambiguous_dna_by_id[2]</span>
<span class="n">mito_table</span> <span class="o">=</span> <span class="n">CodonTable</span><span class="o">.</span><span class="n">unambiguous_dna_by_name</span><span class="p">[</span><span class="s2">&quot;Vertebrate Mitochondrial&quot;</span><span class="p">]</span>


<span class="nb">print</span><span class="p">(</span><span class="n">standard_table</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mito_table</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="MutableSeq">
<h3>MutableSeq<a class="headerlink" href="#MutableSeq" title="Permalink to this headline">¶</a></h3>
<p>Seq objects are <strong>immutable and therefore we cannot change their value</strong>. If we need to do so, we need to convert them into <code class="docutils literal notranslate"><span class="pre">MutableSeq</span></code> objects to avoid a runtime <strong>TypeError</strong> that will occur when we try to change a Seq. Seq to MutableSeq conversion can be done by:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Seq.tomutable()</span></code> to convert a Seq to a MutableSeq.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">MutableSeq.toseq()</span></code> to convert a MutableSeq to a Seq.</p></li>
</ol>
<p>This will allow us to assign the value of characters within the MutableSeq with <code class="docutils literal notranslate"><span class="pre">MutableSeq[ind]</span> <span class="pre">=</span> <span class="pre">newValue</span></code>. Note also that methods like complement change the actual object if applied to a MutableSeq:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">IUPAC</span>

<span class="n">coding_dna</span> <span class="o">=</span> <span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG&quot;</span><span class="p">,</span> <span class="n">IUPAC</span><span class="o">.</span><span class="n">unambiguous_dna</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Coding DNA:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">coding_dna</span><span class="p">)</span>
<span class="n">mut_dna</span> <span class="o">=</span> <span class="n">coding_dna</span><span class="o">.</span><span class="n">tomutable</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Mutable DNA:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mut_dna</span><span class="p">)</span>
<span class="n">mut_dna</span><span class="p">[</span><span class="mi">10</span><span class="p">:</span><span class="mi">20</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;TATATATATA&quot;</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Mutable DNA now:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mut_dna</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Rev comp (Seq not changed!):&quot;</span><span class="p">)</span>
<span class="n">coding_dna</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">coding_dna</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Rev comp (MutableSeq changed!):&quot;</span><span class="p">)</span>
<span class="n">mut_dna</span><span class="o">.</span><span class="n">reverse_complement</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mut_dna</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="Sequence-annotation">
<h2>Sequence annotation<a class="headerlink" href="#Sequence-annotation" title="Permalink to this headline">¶</a></h2>
<p>Sequence records are annotations associated to sequences. In biopython there is a <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> object to store this information. <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> are part of the <code class="docutils literal notranslate"><span class="pre">Bio.SeqRecord</span></code> module which therefore needs to be imported before being used.</p>
<p>The <strong>SeqRecord</strong> class has the following attributes:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.seq</span></code> : the sequence (the Seq object)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.id</span></code> : the identifier of the sequence, typically an accession number</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.name</span></code> : a “common” name or identifier sometimes identical to the accession number</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.description</span></code> : a human readable description of the sequence</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.letter_annotations</span></code> : a per letter annotation using a restricted dictionary (e.g. quality)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.annotations</span></code> : a dictionary of unstructured annotation (e.g. organism, publications,…)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.features</span></code> : a list of SeqFeature objects with more structured information (e.g. genes pos).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord.dbxrefs</span></code> : a list of database cross references.</p></li>
</ol>
<p>Information on SeqFeature objects can be found <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec38">here</a>.</p>
<p>Although you can manually create a SeqRecord, normally you will access SeqRecords populated by methods like SeqIO, <a class="reference external" href="http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec33">here</a> you can find how to manually create a SeqRecord.</p>
</div>
<div class="section" id="Sequence-I/O">
<h2>Sequence I/O<a class="headerlink" href="#Sequence-I/O" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code> module aims to provide a simple way to work with several different sequence file formats. Detailed information on the module can be found <a class="reference external" href="http://biopython.org/wiki/SeqIO">here</a>.</p>
<p><strong>Example:</strong></p>
<p>Read a fasta file <a class="reference external" href="file_samples/NC_005816.fna">NC005816.fna</a> containing the whole sequence for Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1 and retrieve some information about the sequence.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>

<span class="n">record</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/NC_005816.fna&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sequence [first 30 bases]:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">30</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The id:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The description:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;The record is a: &quot;</span><span class="p">,</span> <span class="nb">type</span><span class="p">(</span><span class="n">record</span><span class="p">))</span>

</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
ID: gi|45478711|ref|NC_005816.1|
Name: gi|45478711|ref|NC_005816.1|
Description: gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence
Number of features: 0
Seq(&#39;TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG&#39;)

Sequence [first 30 bases]:
TGTAACGAACGGTGCAATAGTGATCCACAC

The id:
gi|45478711|ref|NC_005816.1|

The description:
gi|45478711|ref|NC_005816.1| Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence

The record is a:  &lt;class &#39;Bio.SeqRecord.SeqRecord&#39;&gt;
</pre></div></div>
</div>
<div class="admonition warning">
<p><strong>WARNING:</strong> When dealing with very large FASTA or FASTQ files, the overhead of working with all these objects can make scripts too slow. In this case <code class="docutils literal notranslate"><span class="pre">SimpleFastaParser</span></code> and <code class="docutils literal notranslate"><span class="pre">FastqGeneralIterator</span></code> parsers might be better as they return just a tuple of strings for each record.</p>
</div>
<p>For example, to use <code class="docutils literal notranslate"><span class="pre">SimpleFastaParser</span></code>:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.SeqIO.FastaIO</span> <span class="kn">import</span> <span class="n">SimpleFastaParser</span>


<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;file_samples/NC_005816.fna&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">in_handle</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">title</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">SimpleFastaParser</span><span class="p">(</span><span class="n">in_handle</span><span class="p">):</span>
        <span class="n">s</span> <span class="o">=</span> <span class="n">seq</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">50</span><span class="p">]</span><span class="o">+</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="n">seq</span><span class="p">[</span><span class="mi">51</span><span class="p">:</span><span class="mi">101</span><span class="p">]</span><span class="o">+</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">...&quot;</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Header:</span><span class="se">\n</span><span class="si">{}</span><span class="se">\n</span><span class="s2">Sequence:</span><span class="se">\n</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">title</span><span class="p">,</span><span class="n">s</span><span class="p">))</span>

<span class="c1">#Another example:</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;1st&quot;</span><span class="p">,</span><span class="s2">&quot;2nd&quot;</span><span class="p">,</span><span class="s2">&quot;3rd&quot;</span><span class="p">]</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&quot;file_samples/contigs82.fasta&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">cont_handle</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">:</span>
        <span class="n">ID</span><span class="p">,</span> <span class="n">seq</span>  <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">SimpleFastaParser</span><span class="p">(</span><span class="n">cont_handle</span><span class="p">))</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="s2">&quot;entry:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">ID</span><span class="p">,</span> <span class="s2">&quot; has size &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">seq</span><span class="p">[:</span><span class="mi">50</span><span class="p">]</span><span class="o">+</span><span class="s2">&quot;...&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

</pre></div>
</div>
</div>
<div class="section" id="BioSeqIO.parse()">
<h3>BioSeqIO.parse()<a class="headerlink" href="#BioSeqIO.parse()" title="Permalink to this headline">¶</a></h3>
<p>The method <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.parse</span></code> is used to parse some sequence data into a <strong>SeqRecord iterator</strong>. In particular, the basic syntax is:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>SeqRecordIterator = Bio.SeqIO.parse(filename, file_format)
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">filename</span></code> is typically an open handle to a file and <code class="docutils literal notranslate"><span class="pre">file_format</span></code> is a lower case string describing the file format. Possible options include <strong>fasta, fastq-illumina, abi, ace, clustal…</strong> all the options are available <a class="reference external" href="http://biopython.org/wiki/SeqIO">here</a>.</p>
<p><strong>Example:</strong> Let’s read the first 10 entries in the .fasta file <a class="reference external" href="file_samples/contigs82.fasta">contigs82.fasta</a> printing off the length of the sequence and the first 50 bases of each sequence followed by “…”.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="n">cnt</span> <span class="o">=</span> <span class="mi">0</span>
<span class="k">for</span> <span class="n">seq_record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;file_samples/contigs82.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">):</span>
    <span class="k">if</span><span class="p">(</span><span class="n">cnt</span> <span class="o">==</span> <span class="mi">10</span><span class="p">):</span>
        <span class="k">break</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Seq </span><span class="si">{}</span><span class="s2"> has length </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq_record</span><span class="p">)))</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">seq_record</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">50</span><span class="p">]</span><span class="o">+</span><span class="s2">&quot;...&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">cnt</span> <span class="o">+=</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
<p>Note that <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.parse</span></code> returns an iterator, therefore it is possible to manually fetch one SeqRecord after the other with the <code class="docutils literal notranslate"><span class="pre">next(iterator)</span></code> method.</p>
<p><strong>Example:</strong> Let’s get the first 3 entries of the .fasta file <a class="reference external" href="file_samples/contigs82.fasta">contigs82.fasta</a> printing off the length of the sequence and the first 50 bases of each sequence followed by “…”.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>

<span class="n">seqIterator</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;file_samples/contigs82.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

<span class="n">labels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;1st&quot;</span><span class="p">,</span><span class="s2">&quot;2nd&quot;</span><span class="p">,</span><span class="s2">&quot;3rd&quot;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">labels</span><span class="p">:</span>
    <span class="n">seqRec</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">seqIterator</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">l</span><span class="p">,</span> <span class="s2">&quot;entry:&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">seqRec</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot; has size &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqRec</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">seqRec</span><span class="o">.</span><span class="n">seq</span><span class="p">[:</span><span class="mi">50</span><span class="p">]</span><span class="o">+</span><span class="s2">&quot;...&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

</pre></div>
</div>
</div>
</div>
<div class="section" id="Sequences-as-dictionaries">
<h3>Sequences as dictionaries<a class="headerlink" href="#Sequences-as-dictionaries" title="Permalink to this headline">¶</a></h3>
<p>The module <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO</span></code> also has three different ways to allow random access to elements:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.to_dict(file_handle/iterator)</span></code> : builds a dictionary of all the SeqRecords keeping them in memory and allowing modifications to the records. <strong>This potentially uses a lot of memory but is very fast</strong>;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.index(filename,file_type)</span></code> : builds a sort of read-only dictionary, parses the elements into SeqRecords on demand (i.e. it returns an iterator!). <strong>This method is slower, but more memory efficient</strong>;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.index_db(indexName.idx,filenames,</span> <span class="pre">file_format)</span></code> : builds a read-only dictionary, but stores ids and offsets on a SQLite3 database. <strong>It is slower but uses less memory.</strong></p></li>
</ol>
<p>The input of <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.to_dict(file_handle/iterator)</span></code> is a open file handler or the iterator to SeqRecords (like the output of Bio.SeqIO.parse()).</p>
<p><strong>Example:</strong> Let’s read all the .fasta file <a class="reference external" href="file_samples/contigs82.fasta">contigs82.fasta</a> in memory printing off the entry corresponding to contig “MDC019140.399”.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="n">seqDict</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;file_samples/contigs82.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">))</span>

<span class="nb">print</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">seqDict</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">:</span><span class="mi">5</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of sequences: &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqDict</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">mySeq</span> <span class="o">=</span> <span class="n">seqDict</span><span class="p">[</span><span class="s2">&quot;MDC019140.399&quot;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Description:&quot;</span><span class="p">,</span> <span class="n">mySeq</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mySeq</span><span class="p">)</span>


</pre></div>
</div>
</div>
<p>The input of <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.index(filename,file_type)</span></code> is the file name to read sequences from and the file type (like fasta, genebank,…).</p>
<p><strong>Example:</strong> Let’s read all the .fasta file <a class="reference external" href="files_sample/contigs82.fasta">contigs82.fasta</a> in memory (as an index) printing off the entry corresponding to contig “MDC019140.399”.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="n">seqDict</span> <span class="o">=</span>  <span class="n">SeqIO</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="s2">&quot;file_samples/contigs82.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;SeqDict is an iterator:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seqDict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">seqDict</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of sequences: &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqDict</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">mySeq</span> <span class="o">=</span> <span class="n">seqDict</span><span class="p">[</span><span class="s2">&quot;MDC019140.399&quot;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Description:&quot;</span><span class="p">,</span> <span class="n">mySeq</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mySeq</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The input of <code class="docutils literal notranslate"><span class="pre">Bio.SeqIO.index_db(indexName.idx,filenames,</span> <span class="pre">file_format)</span></code> is the name of the (SQLite3) index to store (.idx), the file (or files) to index and the format of those files like fasta, genebank,…</p>
<p><strong>Example:</strong> Let’s read all the .fasta file <a class="reference external" href="files_sample/contigs82.fasta">contigs82.fasta</a> in memory (as an index) printing off the entry corresponding to contig “MDC019140.399”.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="n">seqDict</span> <span class="o">=</span>  <span class="n">SeqIO</span><span class="o">.</span><span class="n">index_db</span><span class="p">(</span><span class="s2">&quot;file_samples/ctgs.idx&quot;</span><span class="p">,</span><span class="s2">&quot;file_samples/contigs82.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;SeqDict is an iterator:&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seqDict</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">seqDict</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Number of sequences: &quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">seqDict</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">mySeq</span> <span class="o">=</span> <span class="n">seqDict</span><span class="p">[</span><span class="s2">&quot;MDC019140.399&quot;</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Description:&quot;</span><span class="p">,</span> <span class="n">mySeq</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">mySeq</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Note that a file ctgs.idx has been created to store ids and offsets.</p>
</div>
<div class="section" id="Writing-sequence-files">
<h3>Writing sequence files<a class="headerlink" href="#Writing-sequence-files" title="Permalink to this headline">¶</a></h3>
<p>SeqRecords can be written out to files by using</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>N = Bio.SeqIO.write(records,out_filename, file_format)
</pre></div>
</div>
<p>where <strong>records</strong> is a list of the SeqRecords to write, <strong>out_filename</strong> is the string with the filename to write and <strong>file_format</strong> is the format of the file to write. <strong>N</strong> is the number of sequences written.</p>
<div class="admonition warning">
<p><strong>WARNING:</strong> If you write a file that is already present, <code class="docutils literal notranslate"><span class="pre">SeqIO.write</span></code> will just rewrite it without telling you.</p>
</div>
<p><strong>Example:</strong> Write three protein records to file then read the file and print the three sequences.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">generic_protein</span>

<span class="n">rec1</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MMYQQGCFAGGTVLRLAKDLAENNRGARVLVVCSEITAVTFRGPSETHLDSMVGQALFGD&quot;</span> \
<span class="o">+</span><span class="s2">&quot;GAGAVIVGSDPDLSVERPLYELVWTGATLLPDSEGAIDGHLREVGLTFHLLKDVPGLISK&quot;</span> \
<span class="o">+</span><span class="s2">&quot;NIEKSLKEAFTPLGISDWNSTFWIAHPGGPAILDQVEAKLGLKEEKMRATREVLSEYGNM&quot;</span> \
<span class="o">+</span><span class="s2">&quot;SSAC&quot;</span><span class="p">,</span> <span class="n">generic_protein</span><span class="p">),</span>
<span class="nb">id</span><span class="o">=</span><span class="s2">&quot;gi|14150838|gb|AAK54648.1|AF376133_1&quot;</span><span class="p">,</span>
<span class="n">description</span><span class="o">=</span><span class="s2">&quot;chalcone synthase [Cucumis sativus]&quot;</span><span class="p">)</span>

<span class="n">rec2</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;YPDYYFRITNREHKAELKEKFQRMCDKSMIKKRYMYLTEEILKENPSMCEYMAPSLDARQ&quot;</span> \
<span class="o">+</span><span class="s2">&quot;DMVVVEIPKLGKEAAVKAIKEWGQ&quot;</span><span class="p">,</span> <span class="n">generic_protein</span><span class="p">),</span>
<span class="nb">id</span><span class="o">=</span><span class="s2">&quot;gi|13919613|gb|AAK33142.1|&quot;</span><span class="p">,</span>
<span class="n">description</span><span class="o">=</span><span class="s2">&quot;chalcone synthase [Fragaria vesca subsp. bracteata]&quot;</span><span class="p">)</span>

<span class="n">rec3</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;MVTVEEFRRAQCAEGPATVMAIGTATPSNCVDQSTYPDYYFRITNSEHKVELKEKFKRMC&quot;</span> \
<span class="o">+</span><span class="s2">&quot;EKSMIKKRYMHLTEEILKENPNICAYMAPSLDARQDIVVVEVPKLGKEAAQKAIKEWGQP&quot;</span> \
<span class="o">+</span><span class="s2">&quot;KSKITHLVFCTTSGVDMPGCDYQLTKLLGLRPSVKRFMMYQQGCFAGGTVLRMAKDLAEN&quot;</span> \
<span class="o">+</span><span class="s2">&quot;NKGARVLVVCSEITAVTFRGPNDTHLDSLVGQALFGDGAAAVIIGSDPIPEVERPLFELV&quot;</span> \
<span class="o">+</span><span class="s2">&quot;SAAQTLLPDSEGAIDGHLREVGLTFHLLKDVPGLISKNIEKSLVEAFQPLGISDWNSLFW&quot;</span> \
<span class="o">+</span><span class="s2">&quot;IAHPGGPAILDQVELKLGLKQEKLKATRKVLSNYGNMSSACVLFILDEMRKASAKEGLGT&quot;</span> \
<span class="o">+</span><span class="s2">&quot;TGEGLEWGVLFGFGPGLTVETVVLHSVAT&quot;</span><span class="p">,</span> <span class="n">generic_protein</span><span class="p">),</span>
<span class="nb">id</span><span class="o">=</span><span class="s2">&quot;gi|13925890|gb|AAK49457.1|&quot;</span><span class="p">,</span>
<span class="n">description</span><span class="o">=</span><span class="s2">&quot;chalcone synthase [Nicotiana tabacum]&quot;</span><span class="p">)</span>

<span class="n">my_records</span> <span class="o">=</span> <span class="p">[</span><span class="n">rec1</span><span class="p">,</span> <span class="n">rec2</span><span class="p">,</span> <span class="n">rec3</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">rec1</span><span class="p">)</span>

<span class="n">seqWritten</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">my_records</span><span class="p">,</span> <span class="s2">&quot;file_samples/my_proteins.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">seqWritten</span><span class="p">,</span> <span class="s2">&quot;sequences written!&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Reading file&quot;</span><span class="p">)</span>
<span class="n">seqIT</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;file_samples/my_proteins.fa&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

<span class="n">cnt</span> <span class="o">=</span> <span class="mi">1</span>
<span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">seqIT</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Sequence &quot;</span><span class="p">,</span> <span class="n">cnt</span><span class="p">,</span> <span class="s2">&quot;(&quot;</span><span class="p">,</span> <span class="n">s</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="s2">&quot;)&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">s</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">cnt</span> <span class="o">+=</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="Multiple-sequence-alignment">
<h2>Multiple sequence alignment<a class="headerlink" href="#Multiple-sequence-alignment" title="Permalink to this headline">¶</a></h2>
<p><strong>Multiple Sequence Alignments</strong>, as the name says, are a collection of multiple sequences which have been aligned together – usually with the insertion of gap characters, and addition of leading or trailing gaps – such that all the sequence strings are the same length.</p>
<p><img alt="msa" src="_images/msa.gif" /></p>
<p>Such an alignment can be regarded as a matrix of letters, in biopython each row is a <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> object. These alignments are stored in an object <code class="docutils literal notranslate"><span class="pre">MultipleSeqAlignment</span></code></p>
<div class="section" id="Parsing-multiple-sequence-alignments">
<h3>Parsing multiple sequence alignments<a class="headerlink" href="#Parsing-multiple-sequence-alignments" title="Permalink to this headline">¶</a></h3>
<p>The function <code class="docutils literal notranslate"><span class="pre">Bio.AlignIO.parse()</span></code> returns an iterator of <code class="docutils literal notranslate"><span class="pre">MultipleSeqAlignment</span></code> objects. In the frequent case that we have to deal with a single alignment we will have to use the <code class="docutils literal notranslate"><span class="pre">Bio.AlignIO.read()</span></code> function.</p>
<p>The basic syntax of the two functions:</p>
<div class="line-block">
<div class="line"><code class="docutils literal notranslate"><span class="pre">Bio.AlignIO.parse(file_handle,</span> <span class="pre">alignment_format)</span></code></div>
<div class="line"><code class="docutils literal notranslate"><span class="pre">Bio.AlignIO.read(file_handle,</span> <span class="pre">alignment_format)</span></code></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">file_handle</span></code> is the handler to the opened file, while the <code class="docutils literal notranslate"><span class="pre">alignment_format</span></code> is a lower case string with the alignment format (e.g. fasta, clustal, stockholm, mauve, phylip,…). Check <a class="reference external" href="http://biopython.org/wiki/AlignIO">here</a> for all possible formats.</p>
<p>When more than one multiple alignment has to be read we will have to use the <code class="docutils literal notranslate"><span class="pre">parse</span></code> function, for single entries, we would use the function <code class="docutils literal notranslate"><span class="pre">read</span></code>.</p>
<p><strong>Example:</strong> Load and visualize the seed alignment of the <a class="reference external" href="http://pfam.xfam.org/family/Piwi">Piwi (PF02171) family</a> stored in the pfam (stockholm) format. Information on the format can be found <a class="reference external" href="https://en.wikipedia.org/wiki/Stockholm_format">here</a>. The file is available here: <a class="reference external" href="file_samples/PF02171_seed.sth">PF02171_seed.sth</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>

<span class="n">alignments</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/PF02171_seed.sth&quot;</span><span class="p">,</span> <span class="s2">&quot;stockholm&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">)</span>

</pre></div>
</div>
</div>
<p>The code above prints a summary of the information, but each <code class="docutils literal notranslate"><span class="pre">Bio.Align.MultipleSeqAlignment</span></code> is composed of several alignment records (<code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code>) that we can access with all their information looping through the object. Each <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> contains several information like the <strong>ID, Name, Description, Number of features, start, end and sequence</strong>.</p>
<p><strong>Example:</strong></p>
<p>Load the seed alignment of the <a class="reference external" href="http://pfam.xfam.org/family/Piwi">Piwi (PF02171) family</a> stored in the pfam (stockholm) format <a class="reference external" href="file_samples/PF02171_seed.sth">PF02171_seed.sth</a>. For each record print the description, start and the end point, the sequence and the external dbxrefs (if any).</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>

<span class="n">alignments</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/PF02171_seed.sth&quot;</span><span class="p">,</span> <span class="s2">&quot;stockholm&quot;</span><span class="p">)</span>


<span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;start&quot;</span><span class="p">]</span>
    <span class="n">end</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;end&quot;</span><span class="p">]</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">seq</span>
    <span class="n">desc</span> <span class="o">=</span> <span class="n">align</span><span class="o">.</span><span class="n">description</span>
    <span class="n">dbref</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">align</span><span class="o">.</span><span class="n">dbxrefs</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> S:</span><span class="si">{}</span><span class="s2"> E:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">desc</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">))</span>
    <span class="k">if</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dbref</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">dbref</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">seq</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Writing-multiple-alignments">
<h3>Writing multiple alignments<a class="headerlink" href="#Writing-multiple-alignments" title="Permalink to this headline">¶</a></h3>
<p>Biopython provides a function <code class="docutils literal notranslate"><span class="pre">Bio.AlignIO.write()</span></code> to write alignments to file. The basic syntax is:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>N = Bio.AlignIO.write(alignments,outfile,file_format)
</pre></div>
</div>
<p>where <code class="docutils literal notranslate"><span class="pre">alignments</span></code> are a MultipleSeqAlignment object with the alignments to write to the output file with name <code class="docutils literal notranslate"><span class="pre">outfile</span></code> that has format <code class="docutils literal notranslate"><span class="pre">file_format</span></code> (a low case string with the file format). N is the number of entries written to the file.</p>
<p><strong>Example:</strong> Let’s create some multiple sequence alignments and store them in <a class="reference external" href="http://www.phylo.org/tools/obsolete/phylip.html">phylip</a> format. The output file is present here <a class="reference external" href="file_samples/my_malign.phy">my_malign.phy</a>. Finally, read and print the content of the file.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>

<span></span><span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="kn">import</span> <span class="n">generic_dna</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="kn">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">Bio.Align</span> <span class="kn">import</span> <span class="n">MultipleSeqAlignment</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>

<span class="n">align1</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTGCTAGCTAG&quot;</span><span class="p">,</span> <span class="n">generic_dna</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Alpha&quot;</span><span class="p">),</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACT-CTAGCTAG&quot;</span><span class="p">,</span> <span class="n">generic_dna</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Beta&quot;</span><span class="p">),</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTGCTAGDTAG&quot;</span><span class="p">,</span> <span class="n">generic_dna</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Gamma&quot;</span><span class="p">),</span>
<span class="p">])</span>
<span class="n">align2</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GTCAGC-AG&quot;</span><span class="p">,</span> <span class="n">generic_dna</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Delta&quot;</span><span class="p">),</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GACAGCTAG&quot;</span><span class="p">,</span> <span class="n">generic_dna</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Epsilon&quot;</span><span class="p">),</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;GTCAGCTAG&quot;</span><span class="p">,</span> <span class="n">generic_dna</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Zeta&quot;</span><span class="p">),</span>
<span class="p">])</span>
<span class="n">align3</span> <span class="o">=</span> <span class="n">MultipleSeqAlignment</span><span class="p">([</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTAGTACAGCTG&quot;</span><span class="p">,</span> <span class="n">generic_dna</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Eta&quot;</span><span class="p">),</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;ACTAGTACAGCT-&quot;</span><span class="p">,</span> <span class="n">generic_dna</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Theta&quot;</span><span class="p">),</span>
<span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="s2">&quot;-CTACTACAGGTG&quot;</span><span class="p">,</span> <span class="n">generic_dna</span><span class="p">),</span> <span class="nb">id</span><span class="o">=</span><span class="s2">&quot;Iota&quot;</span><span class="p">),</span>
<span class="p">])</span>

<span class="n">my_alignments</span> <span class="o">=</span> <span class="p">[</span><span class="n">align1</span><span class="p">,</span> <span class="n">align2</span><span class="p">,</span> <span class="n">align3</span><span class="p">]</span>
<span class="n">N</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">my_alignments</span><span class="p">,</span> <span class="s2">&quot;file_samples/my_malign.phy&quot;</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> entries written to the file&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">N</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="n">alignments</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="s2">&quot;file_samples/my_malign.phy&quot;</span><span class="p">,</span> <span class="s2">&quot;phylip&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">)</span>

</pre></div>
</div>
</div>
<p>Note that it is also possible to convert one format into the other (provided that all information needed for the second format is available) by using the function:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Bio.AlignIO.convert(input_file, input_file_format, output_file, output_file_format)
</pre></div>
</div>
<p>basically by passing the input file name and format and output file name and format.</p>
<p>Ex:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>Bio.AlignIO.convert(&quot;PF05371_seed.sth&quot;, &quot;stockholm&quot;, &quot;PF05371_seed.aln&quot;, &quot;clustal&quot;)
</pre></div>
</div>
<p>converts a stockholm file into a clustal.</p>
</div>
<div class="section" id="Manipulating-alignments">
<h3>Manipulating alignments<a class="headerlink" href="#Manipulating-alignments" title="Permalink to this headline">¶</a></h3>
<p>It is possible to slice alignments using the [] operator applied on a <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code>. In particular:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord[i,j]</span></code> returns the jth character of alignment i as a string;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord[:,j]</span></code> returns all the jth characters of the multiple alignment as a string;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord[:,i:j]</span></code> returns a MultipleSeqAlignment with the sub-alignments going for i to j (excluded)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SeqRecord[a:b,i:j]</span></code> similar to 3. but for alignments going from a to b (excluded) only</p></li>
</ol>
<p>One can also combine different portions of the alignment by slicing and adding the slices together with</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>SeqRecord[:,i:j] + SeqRecord[:,t:k]
</pre></div>
</div>
<p><strong>Example:</strong> Load the seed alignment of the <a class="reference external" href="http://pfam.xfam.org/family/Piwi">Piwi (PF02171) family</a> stored in the pfam (stockholm) format <a class="reference external" href="file_samples/PF02171_seed.sth">PF02171_seed.sth</a>. For each record print the description, the first 20 bases and the final 20 bases of the sequence.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>

<span class="n">alignments</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/PF02171_seed.sth&quot;</span><span class="p">,</span> <span class="s2">&quot;stockholm&quot;</span><span class="p">)</span>

<span class="n">edited_aligns</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[:,</span><span class="mi">0</span><span class="p">:</span><span class="mi">20</span><span class="p">]</span> <span class="o">+</span> <span class="n">alignments</span><span class="p">[:,</span><span class="o">-</span><span class="mi">20</span><span class="p">:]</span>


<span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">edited_aligns</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="o">.</span><span class="n">description</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">align</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>

</pre></div>
</div>
</div>
<p><strong>Example:</strong> Convert the seed alignment of the <a class="reference external" href="http://pfam.xfam.org/family/Piwi">Piwi (PF02171) family</a> stored in the pfam (stockholm) format <a class="reference external" href="file_samples/PF02171_seed.sth">PF02171_seed.sth</a> into phylip format. Print some stats on the data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">AlignIO</span>

<span class="n">alignments</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/PF02171_seed.sth&quot;</span><span class="p">,</span> <span class="s2">&quot;stockholm&quot;</span><span class="p">)</span>

<span class="n">out</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">convert</span><span class="p">(</span><span class="s2">&quot;file_samples/PF02171_seed.sth&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;stockholm&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;file_samples/PF02171_seed.aln&quot;</span><span class="p">,</span>
                      <span class="s2">&quot;clustal&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;N. of seq: </span><span class="si">{}</span><span class="se">\n</span><span class="s2">Len of seq: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                                <span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">),</span>
                                                <span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">])))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> multiple alignments converted to phylip&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">out</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Biopython's-pairwise2-alignment">
<h3>Biopython’s pairwise2 alignment<a class="headerlink" href="#Biopython's-pairwise2-alignment" title="Permalink to this headline">¶</a></h3>
<p>Biopython supports running some classic multiple alignment tools like <a class="reference external" href="http://www.sciencedirect.com/science/article/pii/0378111988903307?via%3Dihub">clustal</a>, <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC390337/">MUSCLE</a> but these require to be installed in your system first.</p>
<p>Biopython has its own module to make pairwise alignment. It provides two algorithms: <a class="reference external" href="https://en.wikipedia.org/wiki/Smith%E2%80%93Waterman_algorithm">Smith-Waterman</a> for local alignment and <a class="reference external" href="https://en.wikipedia.org/wiki/Needleman%E2%80%93Wunsch_algorithm">Needleman-Wunsch</a> for global alignment. These methods are implemented in two Biopython functions of the <code class="docutils literal notranslate"><span class="pre">Bio.pairwise2</span></code> module:</p>
<p><code class="docutils literal notranslate"><span class="pre">pairwise2.align.globalxx()</span></code> and <code class="docutils literal notranslate"><span class="pre">pairwise2.align.localxx()</span></code>.</p>
<p>The basic syntax of the two methods is the same:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>aligns = pairwise2.align.globalxx(seq1,seq2)
aligns = pairwise2.align.localxx(seq1,seq2)
</pre></div>
</div>
<p>where seq1 and seq2 are two <code class="docutils literal notranslate"><span class="pre">str</span></code> objects. These methods return a list of alignments (at least one) that have the same <strong>optimal score</strong>. Each alignment is represented as tuples with the following 5 elements in order:</p>
<ol class="arabic simple">
<li><p>The alignment of the first sequence;</p></li>
<li><p>The alignment of the second sequence;</p></li>
<li><p>The alignment score;</p></li>
<li><p>The start of the alignment (for global alignments this is always 0);</p></li>
<li><p>The end of the alignment (for global alignments this is always the length of the alignment).</p></li>
</ol>
<p>The <strong>xx</strong> in <code class="docutils literal notranslate"><span class="pre">globalxx</span></code> or <code class="docutils literal notranslate"><span class="pre">localxx</span></code> encodes the penalties and scores for gaps and matches (and mismatches) of the alignment. The first letter is <strong>the score for a match</strong> the second letter is <strong>the penalty for a gap</strong>:</p>
<p>Match parameters can be:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">x</span></code> : means that a match scores 1 a mismatch 0;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">m</span></code> : the match and mismatch score are passed as additional params after the sequence (es. <code class="docutils literal notranslate"><span class="pre">aligns</span> <span class="pre">=</span> <span class="pre">pairwise2.align.globalmx(seq1,seq2,</span> <span class="pre">1,</span> <span class="pre">-1)</span></code> to set 1 as match score and -1 as mismatch penalty.</p></li>
</ul>
<p>Gap parameters can be:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">x</span></code> : gap penalty is 0;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">s</span></code> : same gap open and gap extend penalties for the 2 sequences (passed as additional params after seqs).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">d</span></code> : different gap open and gap extend penalties for the 2 seqs (additional params after the seqs).</p></li>
</ul>
<p><strong>Example:</strong></p>
<p>Let’s perform the alignment of the two sequences “ACCGTTATATAGGCCA” and “ACGTACTAGTATAGGCCA”</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">pairwise2</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>


<span class="n">alignments</span> <span class="o">=</span> <span class="n">pairwise2</span><span class="o">.</span><span class="n">align</span><span class="o">.</span><span class="n">globalxx</span><span class="p">(</span><span class="s2">&quot;ACCGTTATATAGGCCA&quot;</span><span class="p">,</span>
                                      <span class="s2">&quot;ACGTACTAGTATAGGCCA&quot;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">alignments</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Looping through aligns&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score: </span><span class="si">{}</span><span class="s2">, Start: </span><span class="si">{}</span><span class="s2">, End: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                                                     <span class="n">align</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span>
                                                     <span class="n">align</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

<span class="n">alignments</span> <span class="o">=</span> <span class="n">pairwise2</span><span class="o">.</span><span class="n">align</span><span class="o">.</span><span class="n">globalms</span><span class="p">(</span><span class="s2">&quot;ACCGTTATATAGGCCA&quot;</span><span class="p">,</span>
                                      <span class="s2">&quot;ACGTACTAGTATAGGCCA&quot;</span><span class="p">,</span>
                                      <span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="o">-</span><span class="mf">0.5</span><span class="p">,</span><span class="o">-</span><span class="mf">0.2</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Match: 1, Mismatch: -1, Gap open: -0.5, Gap extend: -0.2&quot;</span><span class="p">)</span>
<span class="k">for</span> <span class="n">align</span> <span class="ow">in</span> <span class="n">alignments</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score: </span><span class="si">{}</span><span class="s2">, Start: </span><span class="si">{}</span><span class="s2">, End: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                                                 <span class="n">align</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span>
                                                 <span class="n">align</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

</pre></div>
</div>
</div>
<p>It is also possible to specify a substitution matrix like the <a class="reference external" href="https://en.wikipedia.org/wiki/BLOSUM">BLOSUM</a> to improve the alignment of protein sequences. Substitution matrices like the <strong>blosum62</strong> are present in Biopython and before using it we need to import it with <code class="docutils literal notranslate"><span class="pre">from</span> <span class="pre">Bio.SubsMat.MatrixInfo</span> <span class="pre">import</span> <span class="pre">blosum62</span></code>.</p>
<p>You can find all the substitution matrices available <a class="reference external" href="https://biopython.org/docs/1.75/api/Bio.SubsMat.MatrixInfo.html">here</a>.</p>
<p><strong>Example:</strong> Align the alpha and beta subunits of the human hemoglobin protein stored in <a class="reference external" href="file_samples/hem_alpha.fasta">hem_alpha.fasta</a> and <a class="reference external" href="file_samples/hem_beta.fasta">hem_beta.fasta</a> with globalxx and with a blosum62 substitution matrix and gap open -10 and gap extension -0.5.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">pairwise2</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.pairwise2</span> <span class="kn">import</span> <span class="n">format_alignment</span>
<span class="kn">from</span> <span class="nn">Bio.SubsMat.MatrixInfo</span> <span class="kn">import</span> <span class="n">blosum62</span>



<span class="n">seq1</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/hem_alpha.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
<span class="n">seq2</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;file_samples/hem_beta.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>


<span class="n">alignments</span> <span class="o">=</span> <span class="n">pairwise2</span><span class="o">.</span><span class="n">align</span><span class="o">.</span><span class="n">globalxx</span><span class="p">(</span><span class="n">seq1</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span><span class="n">seq2</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">),</span> <span class="s2">&quot;optimal alignments&quot;</span><span class="p">)</span>

<span class="n">align</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score: </span><span class="si">{}</span><span class="s2">, Start: </span><span class="si">{}</span><span class="s2">, End: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">align</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">align</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;&quot;</span><span class="p">)</span>

<span class="n">alignments</span> <span class="o">=</span> <span class="n">pairwise2</span><span class="o">.</span><span class="n">align</span><span class="o">.</span><span class="n">globalds</span><span class="p">(</span><span class="n">seq1</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="n">seq2</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="n">blosum62</span><span class="p">,</span> <span class="o">-</span><span class="mi">10</span><span class="p">,</span> <span class="o">-</span><span class="mf">0.5</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">alignments</span><span class="p">),</span> <span class="s2">&quot;optimal alignments&quot;</span><span class="p">)</span>

<span class="n">align</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Score: </span><span class="si">{}</span><span class="s2">, Start: </span><span class="si">{}</span><span class="s2">, End: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">align</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">align</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">align</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span>

</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="Exercises">
<h2>Exercises<a class="headerlink" href="#Exercises" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><p>Write a python function that reads a genebank file given in input and prints off the following information:</p>
<ol class="arabic simple">
<li><p>Identifier, name and description;</p></li>
<li><p>The first 100 characters of the sequence;</p></li>
<li><p>Number of external references (dbxrefs) and ids of the external refs.</p></li>
<li><p>The name of the organism (hint: check the annotations dictionary at the key “organism”)</p></li>
<li><p>Retrieve and print all (if any) associated publications (hint: annotation dictionary, key:”references”)</p></li>
<li><p>Retrieve and print all the locations of “CDS” features of the sequence (hint: check the features )</p></li>
</ol>
</li>
</ol>
<p>Hint: go back and check the details of the <code class="docutils literal notranslate"><span class="pre">SeqRecord</span></code> object.</p>
<p>Test the program downloading some files from genebank like <a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/NC_005816.1/">this</a> or <a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/AB026718.1">this</a></p>
<div class="tggle" onclick="toggleVisibility('ex1');"><p>Show/Hide Solution</p>
</div><div id="ex1" style="display:none;"></div><ol class="arabic simple" start="2">
<li><p>Write a python program that loads a pfam file (stockholm format .sth) and reports for each record of the alignment:</p>
<ol class="arabic simple">
<li><p>the id of the entry</p></li>
<li><p>the start and end points</p></li>
<li><p>the number of gaps and the % of gaps on the total length of the alignment</p></li>
<li><p>the number of external database references (dbxrefs), and the first 3 external references comma separated (hint: use join).</p></li>
</ol>
</li>
</ol>
<p>Print these information to the screen. Finally, write this information in a tab separated file (.tsv) having the following format: <code class="docutils literal notranslate"><span class="pre">#ID\tstart\tend\tnum_gaps\tpercentage_gaps\tdbxrefs</span></code>.</p>
<p>You can test the program using the <a class="reference external" href="https://en.wikipedia.org/wiki/TP53">TP53</a> family file <a class="reference external" href="file_samples/PF00870_seed.sth">PF00870_seed.sth</a>.</p>
<div class="tggle" onclick="toggleVisibility('ex2');"><p>Show/Hide Solution</p>
</div><div id="ex2" style="display:none;"></div><ol class="arabic simple" start="3">
<li><p>Given a multiple sequence alignment stored in “phylip” format, write three methods: <em>readAlignment</em> that reads the input file and prints the number of sequences present, <em>printAlignments(alignments)</em> that prints the alignments to the screen and <em>computeConsensus(alignments,minFrequency)</em> that creates the consensus of all the alignments. MinFrequency is the minimum frequency (that has to be &gt; 0.5) to keep a base in the consensus. “?” is put put if frequency &lt; minFrequency for all possible
bases.</p></li>
</ol>
<p>Ex. if alignments are:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>ATC-G
AAC-G
AAG-G
ATCGT
</pre></div>
</div>
<p>computeConsensus(alignments,0.6) is:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>A?C-G
</pre></div>
</div>
<p>Test the script with the file <a class="reference external" href="file_samples/alpha-globin.phy">alpha-globin.phy</a>.</p>
<div class="tggle" onclick="toggleVisibility('ex3a');"><p>Show/Hide Solution</p>
</div><div id="ex3a" style="display:none;"></div><ol class="arabic simple" start="4">
<li><p>Load the contigs present in the <a class="reference external" href="file_samples/filtered_contigs.fasta">filtered_contigs.fasta</a> file and translate each DNA sequence into the corresponding protein. Count the number of stop codons (i.e. *) for each sequence and print them to the user (e.g. MDC020656.85 51). Finally, write the translated proteins in another .fasta file (e.g. filtered_contigs_translated.fasta).</p></li>
</ol>
<div class="tggle" onclick="toggleVisibility('ex3');"><p>Show/Hide Solution</p>
</div><div id="ex3" style="display:none;"></div><ol class="arabic simple" start="5">
<li><p>Write the following three functions:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">countMatches(s1,s2)</span></code> that takes two sequences aligned with pairwise2 local or global of the same length and returns how many elements match;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">countMismatches(s1,s2)</span></code> that takes two sequences aligned with pairwise2 local or global of the same length and returns how many elements are different (i.e. they are not gaps but the character is different);</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">countGapOpens(s1,s2)</span></code> that takes two sequences aligned with pairwise2 local or global of the same length and returns the number of gap opened in the alignment;</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">countGapExtensions(s1,s2)</span></code> that takes two sequences aligned with pairwise2 local or global of the same length and returns the number of gap extensions in the alignment</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">getScore(s1,s2,</span> <span class="pre">matchScore,</span> <span class="pre">mismatchPenalty,</span> <span class="pre">gapOpenPenalty,</span> <span class="pre">gapExtensionPenalty)</span></code> and returns the score of the alignment of s1 and s2 (aligned with pairwise2 as before) given the scores and penalties in input.</p></li>
</ol>
</li>
</ol>
<p>Finally, align the sequences of the <a class="reference external" href="https://en.wikipedia.org/wiki/Interleukin_12">Interleukin-12</a> chain A (let’s call it s1) <a class="reference external" href="file_samples/IL12A.fasta">IL12A.fasta</a> and B <a class="reference external" href="file_samples/IL12B.fasta">IL12B.fasta</a> (let’s call it s2) with the following parameters:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>pairwise2.align.globalxx(s1,s2)
pairwise2.align.globalms(s1,s2,,1,-0.1,-1,-0.1)
pairwise2.align.globalxs(s1, s2,1,-0.5,-5,-0.5)
pairwise2.align.globalxd(s1, s2,1,-0.5,-5,-0.5,-5,-5)
</pre></div>
</div>
<p>and check the score as computed from pairwise2 and from your functions (use the first of all the alignments to check). Note the effect of changing parameters.</p>
<div class="tggle" onclick="toggleVisibility('ex4');"><p>Show/Hide Solution</p>
</div><div id="ex4" style="display:none;"></div></div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright # 2020 - 2020, Luca Bianco

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>